src/cargo/core/resolver/dep_cache.rs

```diff
@@ -96,7 +96,11 @@ impl<'a> RegistryQueryer<'a> {
     /// any candidates are returned which match an override then the override is
     /// applied by performing a second query for what the override should
     /// return.
-    pub fn query(&mut self, dep: &Dependency) -> Poll<CargoResult<Rc<Vec<Summary>>>> {
+    pub fn query(
+        &mut self,
+        dep: &Dependency,
+        first_minimal_version: bool,
+    ) -> Poll<CargoResult<Rc<Vec<Summary>>>> {
         if let Some(out) = self.registry_cache.get(dep).cloned() {
```

Comment:
Trying to think whether we need to add `first_minimal_version` to the cache key. I think we do, as it's perfectly reasonable to have a transitive and a non-transitive `Dependency` but where this query returns different results.

But either adding it to the cash key or adding a comment explaining why it's not needed would be appreciated.

Reply:
I went with your other comment about `parent.is_none()` though I hesitate about that because that is knowledge outside of the resolver.

Reply:
There are 2 caches here. `self.registry_cache` has `parent` in is cache key. Thank you for adding a comment.

But `self.registry_cache` appears to me not to keep track of the fact that the result of `query` may differ depending on the value of `first_minimal_version`. If I've missed something please add a comment.

Reply:
`self.registry_cache` is now tracking the `first_minimal_version`